src/lib/image-upload-provider.ts

```diff
@@ -19,58 +20,83 @@ interface CollabProvider {
   }
 }
 
-export const imageUploadProvider = (
-  collabProvider: CollabProvider,
-  t: (error: string) => string,
+interface ImageUploadProviderConfig {
+  collabProvider: CollabProvider
+  t: (error: string) => string
   setUploading: (set: boolean) => void
+}
+
+type FileCallback = (file?: File) => void
+
+export const imageUploadProvider = (
+  config: ImageUploadProviderConfig
 ): Promise<ImageUploadProvider> =>
-  Promise.resolve<ImageUploadProvider>((_event, insertImageFn) => {
-    const inputElement = document.createElement(ElementType.Input)
-
-    inputElement.setAttribute('accept', 'image/*')
-
-    inputElement.type = InputType.File
-
-    inputElement.dispatchEvent(new MouseEvent(EventType.Click))
-
-    inputElement.addEventListener(EventType.Change, () => {
-      const reader = new window.FileReader()
-      const file = inputElement.files?.[0]
-
-      if (!file) throw Error(Errors.NoFileFoundAfterInput)
-
-      reader.readAsArrayBuffer(file)
-
-      reader.onloadend = async (): Promise<void> => {
-        const processedFile = processFile(reader.result)
-
-        if (!processedFile) throw Error(Errors.FileNotProcessable)
-
-        try {
-          setUploading(true)
-
-          const {
-            data: { id: src }
-          } = await collabProvider.serviceClient.postImage(
-            file.name,
-            collabProvider.config.noteId,
-            processedFile,
-            file.type
-          )
-
-          insertImageFn({ src })
-        } catch (error) {
-          Alerter.error(
-            t(
-              isCozyStackError(error)
-                ? `Error.${error.status}`
-                : 'Error.unknown_error'
-            ),
-            { duration: TOASTER_DURATION }
-          )
-        } finally {
-          setUploading(false)
-        }
-      }
-    })
-  })
+  Promise.resolve<ImageUploadProvider>((event, insertImageFn) =>
+    handleEvent(event, file => uploadImage(config, insertImageFn, file))
+  )
+
+const handleEvent = (event: Event | undefined, cb: FileCallback): void => {
+  if (event === undefined) return handleClick(cb)
+  if (event.type === EventType.Drop) return handleDrop(event as DragEvent, cb)
+}
+
+const handleClick = (cb: FileCallback): void => {
+  const inputElement = document.createElement(ElementType.Input)
+
+  inputElement.setAttribute('accept', 'image/*')
+
+  inputElement.type = InputType.File
+
+  inputElement.dispatchEvent(new MouseEvent(EventType.Click))
+
+  inputElement.addEventListener(EventType.Change, () =>
+    cb(inputElement.files?.[0])
+  )
+}
+
+const handleDrop = (event: DragEvent, cb: FileCallback): void =>
+  cb(event.dataTransfer?.files?.[0])
```

Comment:
Should not you check that the files[0] is really a file (and not a folder) and specially an image?

Reply:
Well, I'm letting the back-end do that, figured it would be simpler as we have a toaster in case the file is not accepted